src/pipeline.rs

```diff
@@ -0,0 +1,325 @@
+
+    #[cfg(feature = "aio")]
+    async fn execute_transaction_async<C>(&self, con: &mut C) -> RedisResult<Value>
+    where
+        C: crate::aio::ConnectionLike,
+    {
+        let mut resp = con
+            .req_packed_commands(self, self.commands.len() + 1, 1)
+            .await?;
+        match resp.pop() {
+            Some(Value::Nil) => Ok(Value::Nil),
+            Some(Value::Bulk(items)) => Ok(self.make_pipeline_results(items)),
+            _ => Err((
+                ErrorKind::ResponseError,
+                "Invalid response when parsing multi response",
+            )
+                .into()),
+        }
+    }
+
+    /// Async version of `query`.
+    #[inline]
+    #[cfg(feature = "aio")]
+    pub async fn query_async<C, T: FromRedisValue>(&self, con: &mut C) -> RedisResult<T>
+    where
+        C: crate::aio::ConnectionLike,
+    {
+        let v = if self.commands.is_empty() {
+            return from_redis_value(&Value::Bulk(vec![]));
+        } else if self.transaction_mode {
+            self.execute_transaction_async(con).await?
+        } else {
+            self.execute_pipelined_async(con).await?
+        };
+        from_redis_value(&v)
+    }
+
+    /// This is a shortcut to `query()` that does not return a value and
+    /// will fail the task if the query of the pipeline fails.
+    ///
+    /// This is equivalent to a call of query like this:
+    ///
+    /// ```rust,no_run
+    /// # let client = redis::Client::open("redis://127.0.0.1/").unwrap();
+    /// # let mut con = client.get_connection().unwrap();
+    /// let _ : () = redis::pipe().cmd("PING").query(&mut con).unwrap();
+    /// ```
+    ///
+    /// NOTE: A Pipeline object may be reused after `query()` with all the commands as were inserted
+    ///       to them. In order to clear a Pipeline object with minimal memory released/allocated,
+    ///       it is necessary to call the `clear()` before inserting new commands.
+    #[inline]
+    pub fn execute(&self, con: &mut dyn PipelineConnection<Pipeline = Pipeline>) {
+        self.query::<()>(con).unwrap();
+    }
+}
+
+impl Default for Pipeline {
+    fn default() -> Pipeline {
+        Pipeline::new()
+    }
+}
+
+pub(crate) fn encode_pipeline(cmds: &[Cmd], atomic: bool) -> Vec<u8> {
+    let mut rv = vec![];
+    write_pipeline(&mut rv, cmds, atomic);
+    rv
+}
+
+pub(crate) fn write_pipeline(rv: &mut Vec<u8>, cmds: &[Cmd], atomic: bool) {
+    let cmds_len = cmds.iter().map(cmd_len).sum();
+
+    if atomic {
+        let multi = cmd("MULTI");
+        let exec = cmd("EXEC");
+        rv.reserve(cmd_len(&multi) + cmd_len(&exec) + cmds_len);
+
+        multi.write_packed_command_preallocated(rv);
+        for cmd in cmds {
+            cmd.write_packed_command_preallocated(rv);
+        }
+        exec.write_packed_command_preallocated(rv);
+    } else {
+        rv.reserve(cmds_len);
+
+        for cmd in cmds {
+            cmd.write_packed_command_preallocated(rv);
+        }
+    }
+}
```

Comment:
This adds a lot of bulk to the change set, but I think that it's a worthwhile move.